lib/operations/crud_write.js

```diff
@@ -0,0 +1,282 @@
+'use strict';
+
+const operation = require('./operation');
+const { Aspect, defineAspects, OperationBase } = operation;
+const applyRetryableWrites = require('../utils').applyRetryableWrites;
+const applyWriteConcern = require('../utils').applyWriteConcern;
+const handleCallback = require('../utils').handleCallback;
+const decorateWithCollation = require('../utils').decorateWithCollation;
+const toError = require('../utils').toError;
+
+const INSERT = 'insert';
+const UPDATE = 'update';
+const REMOVE = 'remove';
+
+const RETRYABLE_WIRE_VERSION = 6;
+
+function isSingleServer(topology) {
+  if (topology.type && topology.type === 'server') {
+    return true;
+  }
+
+  if (topology.description && topology.description.type === 'Single') {
+    return true;
+  }
+
+  return false;
+}
+
+function isRetryabilitySupported(topology) {
+  if (isSingleServer(topology)) {
+    return false;
+  }
+
+  const maxWireVersion = topology.lastIsMaster().maxWireVersion;
+  if (maxWireVersion < RETRYABLE_WIRE_VERSION) {
+    return false;
+  }
+
+  if (!topology.logicalSessionTimeoutMinutes) {
+    return false;
+  }
+
+  return true;
+}
+
+class CrudWriteOperation extends OperationBase {
+  constructor(model) {
+    super(model.options);
+    this.model = model;
+  }
+
+  canRetry(server, topology) {
+    return (
+      this.options.retryWrites &&
+      this.session &&
+      !this.session.inTransaction() &&
+      isRetryabilitySupported(topology) &&
+      this.model.canRetry()
+    );
+  }
+
+  enableRetry() {
+    this.model.options.willRetryWrite = true;
+    this.session.incrementTransactionNumber();
+  }
+
+  execute(server, callback) {
+    return this.model.execute(server, this.options, callback);
+  }
+}
+
+class CrudModel {
+  constructor(collection, options) {
+    options = Object.assign({}, options);
```

Comment:
I don't think we need to do this here since we do `Object.assign({}, options)` before creating the model in each operation in `collection`.